llmfoundry/callbacks/loss_perp_v_len_callback.py

```diff
@@ -0,0 +1,351 @@
+                )
+
+            if labels.shape[1] != logits.shape[1]:
+                raise ValueError(
+                    f'The length of labels, {labels.shape[1]=} does not match the length of logits {logits.shape[1]=}.',
+                )
+
+            labels, logits = self.preprocess_metric_inputs(
+                sequence_id,
+                labels,
+                logits,
+                seq_parallel_world_size,
+                seq_parallel_rank,
+            )
+
+            self.loss_perp_v_len.update(
+                labels,
+                logits,
+                sequence_id,
+            )
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        if state.timestamp.batch.value % self.compute_batch_interval == 0:
+            current_metric_dict = self.loss_perp_v_len.compute()
+            if dist.get_global_rank() == 0:
+                for k, v in current_metric_dict.items():
+                    v = v.tolist()
+                    v.append(
+                        state.timestamp.batch.value,
+                    )  # Add the current batch index as the last column
+                    if k not in self.metric_dict:
+                        self.metric_dict[k] = []
+                    self.metric_dict[k].append(v)
+        if state.timestamp.batch.value % self.log_batch_interval == 0 and dist.get_global_rank(
+        ) == 0:
+            for k, v in self.metric_dict.items():
+                columns = []
+                columns = [
+                    f'context_length_{i}' for i in range(len(v[0]) - 1)
+                ]  # len(v[0]) - 1 because the last column is the batch index
+                columns.append(
+                    'batch_index',
+                )  # Add batch as the last column name
+                logger.log_table(
+                    columns=columns, # type: ignore
+                    rows=v,
+                    name=f'metrics/train/LossPerpVLenTable/{k}',
+                )
+            self.metric_dict = {}
+
+    def preprocess_metric_inputs(
+        self,
+        sequence_id: Optional[torch.Tensor],
+        labels: torch.Tensor,
+        logits: torch.Tensor,
+        seq_parallel_world_size: int,
+        seq_parallel_rank: int,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        del sequence_id, seq_parallel_rank
+        if seq_parallel_world_size > 1:
+            raise ValueError(
+                'LossPerpVsContextLengthLogger does not support sequence parallelism',
+            )
+
+        return labels, logits
+
+
+class LossPerpVLen(Metric):
+
+    full_state_update = False
+
+    def __init__(
+        self,
+        ignore_index: int,
+        dist_sync_on_step: bool = False,
+    ):
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+
+        self.needs_batch = True
+        self.ignore_index = ignore_index
+        self.loss_fn = torch.nn.CrossEntropyLoss(
```

Comment:
Is it possible to use the model loss fn instead of torch.nn.CrossEntropyLoss
does this make senes?

Reply:
Done.